src/modules/fw_pos_control_l1/fw_pos_control_l1_main.cpp

```diff
@@ -1251,7 +1287,103 @@ FixedwingPositionControl::control_position(const math::Vector<2> &current_positi
+		}
+		/* Reset integrators if switching to this mode from a other mode in which posctl was not active */
+		if (_control_mode_current == FW_POSCTRL_MODE_OTHER) {
+			/* reset integrators */
+			if (_mTecs.getEnabled()) {
+				_mTecs.resetIntegrators();
+				_mTecs.resetDerivatives(_airspeed.true_airspeed_m_s);
+			}
+		}
+		_control_mode_current = FW_POSCTRL_MODE_POSITION;
+
+		/* Get demanded airspeed */
+		float altctrl_airspeed = _parameters.airspeed_min +
+					  (_parameters.airspeed_max - _parameters.airspeed_min) *
+					  _manual.z;
+
+		/* Get demanded vertical velocity from pitch control */
+		static bool was_in_deadband = false;
+		if (_manual.x > deadBand) {
+			float pitch = (_manual.x - deadBand) / factor;
+			_hold_alt -= (_parameters.max_climb_rate * dt) * pitch;
+			was_in_deadband = false;
+		} else if (_manual.x < - deadBand) {
+			float pitch = (_manual.x + deadBand) / factor;
+			_hold_alt -= (_parameters.max_sink_rate * dt) * pitch;
+			was_in_deadband = false;
+		} else if (!was_in_deadband) {
+			 /* store altitude at which manual.x was inside deadBand
+			  * The aircraft should immediately try to fly at this altitude
+			  * as this is what the pilot expects when he moves the stick to the center */
+			_hold_alt = _global_pos.alt;
+			was_in_deadband = true;
+		}
+		tecs_update_pitch_throttle(_hold_alt,
+				altctrl_airspeed,
+				eas2tas,
+				math::radians(_parameters.pitch_limit_min),
+				math::radians(_parameters.pitch_limit_max),
+				_parameters.throttle_min,
+				_parameters.throttle_max,
+				_parameters.throttle_cruise,
+				false,
+				math::radians(_parameters.pitch_limit_min),
+				_global_pos.alt,
+				ground_speed,
+				TECS_MODE_NORMAL);
+
+		/* heading control */
+
+		if (fabsf(_manual.y) < 0.01f) {
+			/* heading / roll is zero, lock onto current heading */
+
+			// XXX calculate a waypoint in some distance
+			// and lock on to it
+
+			/* just switched back from non heading-hold to heading hold */
+			if (!_hdg_hold_enabled) {
+				_hdg_hold_enabled = true;
+				_hdg_hold_yaw = _att.yaw;
+
+				get_waypoint_heading_distance(_hdg_hold_yaw, 3000, _hdg_hold_prev_wp, _hdg_hold_curr_wp);
+			}
+
+			/* we have a valid heading hold position, are we too close? */
+			if (get_distance_to_next_waypoint(_global_pos.lat, _global_pos.lon,
+				_hdg_hold_curr_wp.lat, _hdg_hold_curr_wp.lon) < 1000) {
+				get_waypoint_heading_distance(_hdg_hold_yaw, 3000, _hdg_hold_prev_wp, _hdg_hold_curr_wp);
+			}
+
+			math::Vector<2> prev_wp;
+			prev_wp(0) = (float)_hdg_hold_prev_wp.lat;
+			prev_wp(1) = (float)_hdg_hold_prev_wp.lon;
+
+			math::Vector<2> curr_wp;
+			curr_wp(0) = (float)_hdg_hold_curr_wp.lat;
+			curr_wp(1) = (float)_hdg_hold_curr_wp.lon;
+
+			/* populate l1 control setpoint */
+			_l1_control.navigate_waypoints(prev_wp, curr_wp, current_position, ground_speed_2d);
+
+			_att_sp.roll_body = _l1_control.nav_roll();
+			_att_sp.yaw_body = _l1_control.nav_bearing();
+		} else {
+			_hdg_hold_enabled = false;
```

Comment:
@LorenzMeier If I'm not mistaken then we need to set the manual roll command here?